algorithms/Frequent_Pattern_Mining/fpm/tg_fpm.gsql

```diff
@@ -0,0 +1,147 @@
+                      
+                  INT first_element = s_key,
+                  IF curr_length > 1 THEN first_element = @@pattern_lookup.get(p_key).get(0) END,
+                  
+                  IF curr_length < min_length THEN @@new_pattern_lookup += (key -> concat(@@pattern_lookup.get(p_key), s_key))
+                  ELSE 
+                      @@frequent_pattern += frequent_pattern_node(key, cnt*1.0/unify_size, curr_length),
+                      @@pattern_lookup += (key -> concat(@@pattern_lookup.get(p_key), s_key))
+                  END,
+
+                  s.@work_order += key_pair(key, get_hash_without_first_element(key, first_element, curr_length, @@HASH_CONST)),
+                  @@suffix += (get_hash_without_last_element(key, getvid(s)) -> s)
+                END
+              END
+            POST-ACCUM
+              s.@lookup_table.clear(),
+              s.@counter.clear()
+            HAVING s.@work_order.size() != 0;
+    
+    IF items.size() == 0 or @@frequent_pattern.size() == lim THEN BREAK; END;
+    IF curr_length < min_length THEN
+      @@pattern_lookup = @@new_pattern_lookup;
+      @@new_pattern_lookup.clear();
+    END;
+                      
+    ss = SELECT t
+         FROM items:s - (connect_edge) -> unify_vertex_type:t
+         WHERE t.@is_eligible AND t.item_list.size() >= curr_length
+         ACCUM
+            INT suffix_vid = 0,
+            FOREACH wo in s.@work_order DO
+              FOREACH suffix_page in @@suffix.get(wo.suffix_key) DO
+                 suffix_vid = getvid(suffix_page),
+                      
+                 IF is_subset(@@pattern_lookup.get(wo.pattern_key), suffix_vid, t.item_list) THEN
+                   suffix_page.@counter += (get_hash_concat(wo.pattern_key, suffix_vid) -> 1),
+                   suffix_page.@lookup_table += (get_hash_concat(wo.pattern_key, suffix_vid) 
+                                               -> key_pair(wo.pattern_key, suffix_vid)),
+                   t.@superset_counter += 1
+                 END
+              END
+           END
+         POST_ACCUM s.@work_order.clear(),
+                    t.@is_eligible = t.@superset_counter != t.@superset_counter';
+                      
+    @@suffix.clear();
+  END;
+
+  ##### Output #####                    
+  IF spit_to_file THEN
+    f.println("FREQUENCY", "PATTERN_LIST");
+  END;
+
+  IF output_limit != 0 OR spit_to_file THEN
+    WHILE NOT @@frequent_pattern.size() == 0 DO
+      IF (output_limit == -1 or output_limit > 0 and print_count < output_limit) THEN
+        PRINT @@frequent_pattern.top().frequency as support,
+              to_vertex_list(@@pattern_lookup.get(@@frequent_pattern.top().pattern_key)) as pattern;
+                      
+        print_count = print_count + 1;
+      END;
+
+      IF spit_to_file THEN
+        f.println(@@frequent_pattern.top().frequency, @@pattern_lookup.get(@@frequent_pattern.top().pattern_key));   
+      END;
```

Comment:
having `output_limit` for pint result makes sense,
not sure why a user might want to limit the output to file, maybe we should always output the entire result to file?